nucypher/blockchain/eth/domains.py

```diff
@@ -18,38 +24,72 @@ class PolygonChain(ChainInfo, Enum):
     MUMBAI = (80001, "mumbai")
 
 
-class DomainInfo(NamedTuple):
-    name: str
-    eth_chain: EthChain
-    polygon_chain: PolygonChain
+class TACoDomain:
+    def __init__(self, name: str, eth_chain: EthChain, polygon_chain: PolygonChain):
+        self.name = name
+        self.eth_chain = eth_chain
+        self.polygon_chain = polygon_chain
+
+    def __repr__(self) -> str:
+        return f"<TACoDomain {self.name}>"
+
+    def __str__(self) -> str:
+        return self.name
+
+    def __hash__(self) -> int:
+        return hash(self.name)
+
+    def __bytes__(self) -> bytes:
+        return self.name.encode()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, TACoDomain):
+            return False
+        return (
+            self.name == other.name
+            and self.eth_chain == other.eth_chain
+            and self.polygon_chain == other.polygon_chain
+        )
+
+    def __bool__(self) -> bool:
+        return True
 
     @property
     def is_testnet(self) -> bool:
         return self.eth_chain != EthChain.MAINNET
 
 
-class TACoDomain:
-    class Unrecognized(RuntimeError):
-        """Raised when a provided domain name is not recognized."""
+MAINNET = TACoDomain(
+    name="mainnet",
+    eth_chain=EthChain.MAINNET,
+    polygon_chain=PolygonChain.MAINNET,
+)
+
+LYNX = TACoDomain(
+    name="lynx",
+    eth_chain=EthChain.GOERLI,
+    polygon_chain=PolygonChain.MUMBAI,
+)
 
-    MAINNET = DomainInfo("mainnet", EthChain.MAINNET, PolygonChain.MAINNET)
-    LYNX = DomainInfo("lynx", EthChain.GOERLI, PolygonChain.MUMBAI)
-    TAPIR = DomainInfo("tapir", EthChain.SEPOLIA, PolygonChain.MUMBAI)
+TAPIR = TACoDomain(
+    name="tapir",
+    eth_chain=EthChain.SEPOLIA,
+    polygon_chain=PolygonChain.MUMBAI,
+)
 
-    DEFAULT_DOMAIN_NAME: str = MAINNET.name
 
-    SUPPORTED_DOMAINS = [
-        MAINNET,
-        LYNX,
-        TAPIR,
-    ]
+DEFAULT_DOMAIN: TACoDomain = MAINNET
 
-    SUPPORTED_DOMAIN_NAMES = [domain.name for domain in SUPPORTED_DOMAINS]
+SUPPORTED_DOMAINS: Dict[str, TACoDomain] = {
+    str(domain): domain for domain in (MAINNET, LYNX, TAPIR)
+}
 
-    @classmethod
-    def get_domain_info(cls, domain: str) -> DomainInfo:
-        for taco_domain in cls.SUPPORTED_DOMAINS:
-            if taco_domain.name == domain:
-                return taco_domain
 
-        raise cls.Unrecognized(f"{domain} is not a recognized domain.")
+@memoize
+def get_domain(d: Any) -> TACoDomain:
```

Comment:
since I'm type checking on the first line of the function body I'm entertaining the idea that this public function may receive an invalid input that is not a string.